src/server/journal/streamer.cc

```diff
@@ -312,7 +310,7 @@ bool RestoreStreamer::WriteBucket(PrimeTable::bucket_iterator it) {
 void RestoreStreamer::OnDbChange(DbIndex db_index, const DbSlice::ChangeReq& req) {
   DCHECK_EQ(db_index, 0) << "Restore migration only allowed in cluster mode in db0";
 
-  FiberAtomicGuard fg;
+  { std::unique_lock<util::fb2::Mutex> lk(bucket_ser_mu_); }
```

Comment:
why do you put it on its own scope? It will be freed immediately after, which will mean that other `OnDbChange` callbacks could run in parallel, no?

Reply:
I don't think so this is the case. Once it acquires the lock, it will drop it immediately but it won't `preempt`. It will `preempt` potentially when we call `WriteBucket`. All the steps between dropping the lock on line 313 and calling `WriteBucket` are atomic, there is no preemption. However, when we call `WriteBucket` we will reqacquire the lock on the same fiber `before we preempt`. Think of this `{}` as transferring ownership of the lock without explicitly doing it (e.g. `lk.unlock()` before we call `WriteBucket`

Reply:
Some thoughts on what you wrote:
* Is what you wrote also true for `CVCUponInsert()`? I see that it uses a callback. I _think_ it's true, but you should verify if we're counting on that
* I know that "regular" mutexes (i.e. non fiber ones) do not guarantee that, even if the lock is unlocked, it will not preempt to other threads. I don't know if that's true here, but I guess it is. In other words, this may not _guarantee_ a mutex transfer without yielding in between
* I understand that `WriteBucket()` locks this lock, so I don't see any benefit to lock (and immediately release) here as well. Why even do that?
* There might be a subtle bug here: `CVCUponInsert()` might call `WriteBucket()` multiple times (over multiple buckets). If we lose the mutex to some other operation, which touches that other bucket, it might change the underlying structure causing us to lose some changes

I'd recommend adding a `WriteBucketNoLock()` method which will be called here and in `WriteBucket()`. Bonus points if you add `ABSL_LOCKS_EXCLUDED` and `ABSL_LOCKS_REQUIRED` annotations to let the compiler verify we don't create deadlocks

Reply:
I also dont see why we need to lock and release here, as if we lock in WriteBucket this should be enough.
@chakaz Regarding the comment that CVCUponInsert calls WriteBucket multiple times, I think that ok.. because what we care is that we dont break the bucket into multiple pieces but I dont think we should care if CVCUponInsert serializes one bucket we will preempt , serialize onother bucket from different flow and than go back to continue serialize the next buckets from CVCUponInsert. Is that right?

Reply:
@chakaz 

Yes we do not need the lock `OnDbChange` that was a mistake.

> I know that "regular" mutexes (i.e. non fiber ones) do not guarantee that, even if the lock is unlocked, it will not preempt to other threads. I don't know if that's true here, but I guess it is. In other words, this may not guarantee a mutex transfer without yielding in between

Regular mutexes behave differently when notifying other threads to make progress. If this code was running on thread a but wakes up thread b and releases the mutex then it's a race for which thread locks the mutex. This is not true for fibers: 

```
   90 void Mutex::unlock() {                                                                             
   91   detail::FiberInterface* active = detail::FiberActive();                                          
   92                                                                                                    
   93   unique_lock lk(wait_queue_splk_);                                                                
   94   CHECK(owner_ == active);                                                                         
   95   owner_ = nullptr;                                                                                
   96                                                                                                    
   97   wait_queue_.NotifyOne(active);                                                                   
   98 }  
```

And if you follow `NotifyOne` it calls `NotifyImpl` which calls `ActivateOther` 

```
  116   
  117   // Schedules another fiber without switching to it.                                              
  118   // other can belong to another thread.                                                           
  119   void ActivateOther(FiberInterface* other);
  120   
```

The gist here is a) That ActivateOther adds the fiber to the ready queue b) that in our case all the interesting fibers (the fiber that runs OnDbChange && IterateFB) that use this mutex run on the same `thread`, this is both true for SaveStagesController && FullSyncFb for replication (on RDB version they do push on the same serialization channel which is fine since we don’t (yet until phase 2) split individual slots.

+1 for the rest 

@adiholden won't happen, when we release the lock we continue running on the same fiber, we don't switch.

Plz let me know if I missed anything

Reply:
@adiholden 

>  because what we care is that we dont break the bucket into multiple pieces
We do want to break the bucket into multiple pieces, otherwise what's the point ?

Reply:
> We do want to break the bucket into multiple pieces, otherwise what's the point ?

We want to break _entries_, not _buckets_. Breaking on a bucket granularity isn't enough.
Of course the challenge is no to have interleaved entries in the stream.